compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/service/validator/ServiceValidator.java

```diff
@@ -132,6 +138,75 @@ public void validate() {
+            }
+            validateEntityAnnotation(annotationNode);
+        }
+    }
+
+    private AnnotationNode getEntityAnnotationNode(AnnotationSymbol annotationSymbol, String entityName) {
+        Project currentProject = this.context.currentPackage().project();
+        EntityAnnotationFinder entityAnnotationFinder = new EntityAnnotationFinder(this.context.semanticModel(),
+                                                                                   currentProject,
+                                                                                   this.context.moduleId(),
+                                                                                   annotationSymbol, entityName);
+        return entityAnnotationFinder.find().orElse(null);
+    }
+
+    private void validateEntityAnnotation(AnnotationNode entityAnnotation) {
+        if (entityAnnotation.annotValue().isEmpty()) {
+            return;
+        }
+        for (MappingFieldNode fieldNode : entityAnnotation.annotValue().get().fields()) {
+            if (fieldNode.kind() != SPECIFIC_FIELD) {
+                addDiagnostic(CompilationDiagnostic.PROVIDE_KEY_VALUE_PAIR_FOR_ENTITY_ANNOTATION, fieldNode.location());
+                continue;
+            }
+            SpecificFieldNode specificFieldNode = (SpecificFieldNode) fieldNode;
+            Node fieldNameNode = specificFieldNode.fieldName();
+            if (fieldNameNode.kind() != SyntaxKind.IDENTIFIER_TOKEN) {
+                continue;
+            }
+            IdentifierToken fieldNameToken = (IdentifierToken) fieldNameNode;
+            String fieldName = fieldNameToken.text().trim();
+            if (fieldName.equals(KEY)) {
+                validateKeyFieldValue(specificFieldNode);
+            }
+        }
+    }
+
+    private void validateKeyFieldValue(SpecificFieldNode specificFieldNode) {
+        if (specificFieldNode.valueExpr().isEmpty()) {
+            addDiagnostic(CompilationDiagnostic.PROVIDE_A_STRING_LITERAL_OR_AN_ARRAY_OF_STRING_LITERALS_FOR_KEY_FIELD,
+                          specificFieldNode.location(), KEY);
+            return;
+        }
+        ExpressionNode keyFieldExpression = specificFieldNode.valueExpr().get();
+        if (keyFieldExpression.kind() == SyntaxKind.STRING_LITERAL) {
+            return;
+        }
+        if (keyFieldExpression.kind() == SyntaxKind.LIST_CONSTRUCTOR) {
+            for (Node expression : ((ListConstructorExpressionNode) keyFieldExpression).expressions()) {
+                if (expression.kind() != SyntaxKind.STRING_LITERAL) {
+                    addDiagnostic(CompilationDiagnostic.PROVIDE_AN_ARRAY_OF_STRING_LITERALS_FOR_KEY_FIELD,
+                                  expression.location(), KEY);
+                }
+            }
+            return;
+        }
+        addDiagnostic(CompilationDiagnostic.PROVIDE_A_STRING_LITERAL_OR_AN_ARRAY_OF_STRING_LITERALS_FOR_KEY_FIELD,
+                      keyFieldExpression.location(), KEY);
```

Comment:
This function flow looks counter-intuitive. First `if` block adds a diagnostic and returns, while the second if block just returns. We may invert the second condition to add the diagnostic inside that block. 

When it is a list, aren't we doing the same check for list members? Can't we use a separate function to validate each expression and call it inside the loop?

Reply:
If we invert the second 'if' block, it will always result in a warning for a list constructor expression. We could modify that 'if' block to check both kinds (not equals to 'STRING_LITERAL' and 'LIST_CONSTRUCTOR'), move the warning on line 208 inside this block, and then return. This way, both solutions will yield the same result.

Reply:
Then we should check for the list first. 

My understanding is, at the end of the day, what we want to check is whether a given expression is valid or not. So, as per the single responsibility principle, one function should only do that. Then there's the key argument, which can be either an expression or a list. For a list, we should iterate and pass each expression to the aforementioned function. Am I missing something?

Reply:
Got the point now. Separate the concerns of iteration and validation with: [a02f5db](https://github.com/ballerina-platform/module-ballerina-graphql/pull/1579/commits/a02f5dbead0dd96f4a4984827d16674941a774df)